### PR TITLE
Set-DbaDbCompression - Refactoring

### LINKED ISSUE #7864: Set-DbaDbCompression error when multiple databases
### Verified issue does not already exist?

Yes

### What error did you receive?

WARNING: [15:55:11][Set-DbaDbCompression] Compression failed for localhost - [DB1] | Error connecting to [localhost]: Method invocation failed because [System.Management.Automation.PSObject] does not contain a method named
 'op_Addition'.
WARNING: [15:55:11][Set-DbaDbCompression] Compression failed for localhost - [DB2] | Error connecting to [localhost]: Method invocation failed because [System.Management.Automation.PSObject] does not contain a method
named 'op_Addition'.
WARNING: [15:55:11][Set-DbaDbCompression] Compression failed for localhost - [DB3] | Error connecting to [localhost]: Method invocation failed because [System.Management.Automation.PSObject] does not contain a method named
'op_Addition'.
 named 'op_Addition'.
WARNING: [15:55:11][Set-DbaDbCompression] Compression failed for localhost - [DB4] | Error connecting to [localhost]: Method invocation failed because [System.Management.Automation.PSObject] does not contain a method named
'op_Addition'.
WARNING: [15:55:11][Set-DbaDbCompression] Compression failed for localhost - [DB5] | Error connecting to [localhost]: Method invocation failed because [System.Management.Automation.PSObject] does not contain a method named
 'op_Addition'.
WARNING: [15:55:11][Set-DbaDbCompression] Compression failed for localhost - [DB6] | Error connecting to [localhost]: Method invocation failed because [System.Management.Automation.PSObject] does not contain a method named
'op_Addition'.

### Steps to Reproduce

```powershell
# provide your command(s) executed pertaining to dbatools
# please include variable values (redacted or fake if needed) for reference
```
$InputObject = Test-DbaDbCompression -SqlInstance <instancename> -ExcludeDatabase master,model,msdb,tempdb -SqlCredential $cred | where { $_.SizeCurrent.Megabyte -gt 100.0 -and $_.PercentCompression -ge 25.0 }
Set-DbaDbCompression -SqlInstance <instancename> -SqlCredential $cred -InputObject $InputObject

### Are you running the latest release?

Yes

### Other details or mentions

Test-DbaDbCompression completes without error (takes a long time though).
Set-DbaDbCompression throws warnings immediately and quits.

### What PowerShell host was used when producing this error

Windows PowerShell (powershell.exe)

### PowerShell Host Version

Name                           Value
----                           -----
PSVersion                      5.1.17763.2183
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17763.2183
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

### SQL Server Edition and Build number

Microsoft SQL Server 2019 (RTM-CU12) (KB5004524) - 15.0.4153.1 (X64)   Jul 19 2021 15:37:34   Copyright (C) 2019 Microsoft Corporation  Enterprise Edition (64-bit) on Windows Server 2019 Datacenter 10.0 <X64> (Build 17763: ) (Hypervisor) 

### .NET Framework Version

Get-ChildItem 'HKLM:\SOFTWARE\Microsoft\NET Framework Setup\NDP' -recurse |
>> Get-ItemProperty -name Version,Release -EA 0 |
>> Where { $_.PSChildName -match '^(?!S)\p{L}'} |
>> Select PSChildName, Version, Release

PSChildName Version   Release
----------- -------   -------
Client      4.7.03190  461814
Full        4.7.03190  461814
Client      4.0.0.0

[System.Runtime.InteropServices.RuntimeInformation]::get_FrameworkDescription() did not work.
Method invocation failed because [System.Runtime.InteropServices.RuntimeInformation] does not contain a method named 'get_FrameworkDescription'.
At line:1 char:1
+ [System.Runtime.InteropServices.RuntimeInformation]::get_FrameworkDes ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound

### PR DESCRIPTION
**Pull request by andreasjordan**
## Type of Change

 - [x] Bug fix (non-breaking change, fixes #7864  )
 - [ ] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1`)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/dataplat/appveyor-lab ?
 - [ ] Unit test is included
 - [ ] Documentation
 - [ ] Build system

To solve a lot of issues, I went for a big refactoring. I used smaller try-catch blocks to help troubleshooting in the future. Only used try-catch where code could fail.

### CODE REVIEW

**Comment by potatoqualitee** [comment]
@jpomfret please take a look when you have a moment! i will merge once approved 👍🏼

**Comment by jpomfret** [comment]
Hey folks!
Struggling to repro the issue in #7864 - so making it hard to test, will have another go a little later.

**Comment by andreasjordan** [comment]
I could not repor it either. So I just "faked" the $InputObject as the code of Test-DbaDbCompression nicely shows the structure of the PSCustomObject.

File: functions/Set-DbaDbCompression.ps1
Comment (wsmelton) [inline functions/Set-DbaDbCompression.ps1:196]: In practice, Set commands should only output what they changed.

**Comment by jpomfret** [comment]
> I could not repor it either. So I just "faked" the $InputObject as the code of Test-DbaDbCompression nicely shows the structure of the PSCustomObject.

Ah ok - I was heading that direction but ran out of time yesterday. Is it worth adding a test for multiple databases\multiple tables?

**Comment by andreasjordan** [comment]
I can now repo, here is the code:

```
# Part 1: Two databases with two tables each:
$db1 = New-DbaDatabase -SqlInstance sql01 -Name CompressionTest1
$db1.Query("CREATE TABLE dbo.tab1 (col1 CHAR(100) DEFAULT 'A', col2 INT DEFAULt 1)")
1..100 | ForEach-Object { $db1.Query("INSERT INTO dbo.tab1 DEFAULT VALUES") }
$db1.Query("CREATE TABLE dbo.tab2 (col1 CHAR(100) DEFAULT 'B', col2 INT DEFAULt 2)")
1..100 | ForEach-Object { $db1.Query("INSERT INTO dbo.tab2 DEFAULT VALUES") }

$db2 = New-DbaDatabase -SqlInstance sql01 -Name CompressionTest2
$db2.Query("CREATE TABLE dbo.tab1 (col1 CHAR(100) DEFAULT 'A', col2 INT DEFAULt 1)")
1..100 | ForEach-Object { $db2.Query("INSERT INTO dbo.tab1 DEFAULT VALUES") }
$db2.Query("CREATE TABLE dbo.tab2 (col1 CHAR(100) DEFAULT 'B', col2 INT DEFAULt 2)")
1..100 | ForEach-Object { $db2.Query("INSERT INTO dbo.tab2 DEFAULT VALUES") }

$InputObject = Test-DbaDbCompression -SqlInstance sql01 -Database CompressionTest1, CompressionTest2
$InputObject | ForEach-Object { $_.CompressionTypeRecommendation = 'PAGE' ; $_.PercentCompression = 20 }
$InputObject | Format-Table

Set-DbaDbCompression -SqlInstance sql01 -InputObject $InputObject | Format-Table

$null = Remove-DbaDatabase -SqlInstance sql01 -Database CompressionTest1, CompressionTest2 -Confirm:$false
# Problem: Every table is compressed more than once, because or "$results +="



# Part 2: Two databases with only one table each.
$db1 = New-DbaDatabase -SqlInstance sql01 -Name CompressionTest1
$db1.Query("CREATE TABLE dbo.tab1 (col1 CHAR(100) DEFAULT 'A', col2 INT DEFAULt 1)")
1..100 | ForEach-Object { $db1.Query("INSERT INTO dbo.tab1 DEFAULT VALUES") }

$db2 = New-DbaDatabase -SqlInstance sql01 -Name CompressionTest2
$db2.Query("CREATE TABLE dbo.tab1 (col1 CHAR(100) DEFAULT 'A', col2 INT DEFAULt 1)")
1..100 | ForEach-Object { $db2.Query("INSERT INTO dbo.tab1 DEFAULT VALUES") }

$InputObject = Test-DbaDbCompression -SqlInstance sql01 -Database CompressionTest1, CompressionTest2
$InputObject | ForEach-Object { $_.CompressionTypeRecommendation = 'PAGE' ; $_.PercentCompression = 20 }
$InputObject | Format-Table

Set-DbaDbCompression -SqlInstance sql01 -InputObject $InputObject | Format-Table

$null = Remove-DbaDatabase -SqlInstance sql01 -Database CompressionTest1, CompressionTest2 -Confirm:$false
# Problem: Failure like in #8148, because or "$results +="
```

The new code solves both problems.

File: functions/Set-DbaDbCompression.ps1
Comment (andreasjordan) [reply functions/Set-DbaDbCompression.ps1:196]: I agree. What is `.AlreadyProcessed` for - if it is always true in the output?

**Comment by potatoqualitee** [comment]
will merge on @jpomfret approval 💯

**Comment by jpomfret** [comment]
Yeah go ahead - looks ok to me

**Review by jpomfret — APPROVED** [review]

**Comment by potatoqualitee** [comment]
excellent, thank you both 🙇🏼

**Pull request merged**
